### PR TITLE
MariaCraciun1_task2

### PR DESCRIPTION
**Pull request by MariaCraciun1**
Initial Commit

### CODE REVIEW

File: MariaCraciun1_Task2/BoostTest.cpp [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+
+  BOOST_AUTO_TEST_CASE( getEntryTest )
+  {
+      DynamicStringArray obj;
+      obj.addEntry ( "Ana" );
+      BOOST_CHECK ( obj.getEntry(0) -> compare ( "Ana" ) == 0 );
+  }
+
+  BOOST_AUTO_TEST_CASE( getLengthTest )
+  {
```
Comment (maxlig79) [inline MariaCraciun1_Task2/BoostTest.cpp:41]: Is the same test as addEntry

File: MariaCraciun1_Task2/BoostTest.cpp [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+      BOOST_CHECK( obj1.getLength() == obj2.getLength() );
+  }
+
+  BOOST_AUTO_TEST_CASE( defaultConstructorTest )
+  {
+      DynamicStringArray obj;
+      if(obj.getLength()!=0 && obj.getEntry(0) != nullptr )
+      BOOST_FAIL( "Initialization failed" );
+  }
+
```
Comment (maxlig79) [inline MariaCraciun1_Task2/BoostTest.cpp:74]: From the tests you missed getEntry

File: MariaCraciun1_Task2/DynamicStringArray.cpp [hunk 3]
```diff
@@ -0,0 +1,84 @@
+    #include "DynamicStringArray.hpp"
+
+
+    DynamicStringArray::DynamicStringArray(){
+        size=0;
+        dynamicArray=nullptr;
```
Comment (maxlig79) [inline MariaCraciun1_Task2/DynamicStringArray.cpp]: As a suggestion is better to use initializer list

File: MariaCraciun1_Task2/DynamicStringArray.cpp [hunk 4]
```diff
@@ -0,0 +1,84 @@
...
+        }
+     return *this;       
+    }
+    int DynamicStringArray::getLength() const{
+        return size;
+    }
+   
+    void DynamicStringArray::addEntry(const std::string &str){
+        
+        if(size==0){
```
Comment (maxlig79) [inline MariaCraciun1_Task2/DynamicStringArray.cpp]: This is not required, should be handled in the else block

File: MariaCraciun1_Task2/DynamicStringArray.cpp [hunk 5]
```diff
@@ -0,0 +1,84 @@
...
+            dynamicArray=tempArray;
+        }
+    }
+    bool DynamicStringArray::deleteEntry(const std::string &str){
+
+        if(dynamicArray[0]==str && size==1){
+            delete[] dynamicArray;
+            dynamicArray=nullptr;
+            size--;
+            return true;
```
Comment (maxlig79) [inline MariaCraciun1_Task2/DynamicStringArray.cpp]: Please avoid multiple return

File: MariaCraciun1_Task2/DynamicStringArray.cpp [hunk 6]
```diff
@@ -0,0 +1,84 @@
...
+        }
+        else{
+            std::string *tempArray= new std::string[size+1];
+            for(int i=0;i<size;i++){
+                tempArray[i]=dynamicArray[i];
+            }
+            tempArray[size]=str;
+            size++;
+            delete[] dynamicArray;
+            dynamicArray=nullptr;
```
Comment (maxlig79) [inline MariaCraciun1_Task2/DynamicStringArray.cpp]: No need for this if you imediatly assign it.

File: MariaCraciun1_Task2/DynamicStringArray.cpp [hunk 7]
```diff
@@ -0,0 +1,84 @@
...
+        size=obj.getLength();
+        dynamicArray= new std::string[size];
+        for(int i=0;i<size;i++){
+            dynamicArray[i]=obj.dynamicArray[i];
+        }
+    }
+    
+    DynamicStringArray &DynamicStringArray::operator=(const DynamicStringArray &obj){
+        size= obj.getLength();
+        dynamicArray= new std::string[size];
```
Comment (maxlig79) [inline MariaCraciun1_Task2/DynamicStringArray.cpp]: What happen here when you have self assignment?

File: MariaCraciun1_Task2/BoostTest.cpp [hunk 8]
```diff
@@ -0,0 +1,73 @@
...
+BOOST_AUTO_TEST_SUITE( DynamicStringArrayTests)
+
+  BOOST_AUTO_TEST_CASE( addEntryTest ) 
+  {
+      DynamicStringArray obj;
+      obj.addEntry ( "Ana" );
+      obj.addEntry ( "are" );
+      obj.addEntry ( "mere" );
+      BOOST_CHECK( obj.getLength() == 3 );
```
Comment (maxlig79) [inline MariaCraciun1_Task2/BoostTest.cpp]: Maybe here you can use getEntry?

**Review by maxlig79 — CHANGES_REQUESTED** [review]

File: MariaCraciun1_Task2/DynamicStringArray.cpp (hunk 7 above)
Comment (MariaCraciun1) [reply MariaCraciun1_Task2/DynamicStringArray.cpp]: That case was not treated. I have fixed this by putting everything, except the return statement in this if :
  if ( this != &obj )
This way if there is self assignment, the current object will be returned.

File: MariaCraciun1_Task2/DynamicStringArray.cpp [hunk 9]
```diff
@@ -0,0 +1,75 @@
...
+        }
+        std::string *tempArray = new std::string[size-1] ;
+        int j=0 ;
+        for ( int i=0; i<size; i++){
+            if ( dynamicArray[i] != str )
+                tempArray[j] = dynamicArray[i] ;
+                j++ ;
+        }
+        delete[] dynamicArray ; 
+        dynamicArray = nullptr ;
```
Comment (maxlig79) [inline MariaCraciun1_Task2/DynamicStringArray.cpp]: No need to do this since in the next line you will assign it to something new.

**Review by maxlig79 — CHANGES_REQUESTED** [review]

File: MariaCraciun1_Task2/BoostTest.cpp [hunk 10]
```diff
@@ -0,0 +1,75 @@
...
+      BOOST_CHECK ( obj.getEntry(2)->compare("mere") == 0 );
+  }
+
+  BOOST_AUTO_TEST_CASE ( deleteEntryTest )
+  {
+      DynamicStringArray obj;
+      obj.addEntry ( "Ana" );
+      obj.addEntry ( "are" );
+      obj.addEntry ( "mere" );
+      if(obj.deleteEntry ( "mere" ) != true && obj.getLength() !=2 )
```
Comment (nbiro) [inline MariaCraciun1_Task2/BoostTest.cpp:25]: this works, but you could also use BOOST_CHECK_MESSAGE to print a nice error message

File: MariaCraciun1_Task2/DynamicStringArray.cpp [hunk 11]
```diff
@@ -0,0 +1,72 @@
...
+        size++ ;
+        dynamicArray = tempArray ;
+    }
+    
+    bool DynamicStringArray::deleteEntry (const std::string &str) {
+
+        bool status = false;
+        if ( dynamicArray[0] == str && size == 1 ) { 
+            delete[] dynamicArray ;
+            dynamicArray = nullptr ;
```
Comment (maxlig79) [inline MariaCraciun1_Task2/DynamicStringArray.cpp:43]: No need for this

File: MariaCraciun1_Task2/BoostTest.cpp [hunk 12]
```diff
@@ -0,0 +1,75 @@
...
+        BOOST_ERROR ( "This test should fail since the index is out of bounds." ) ;
+  }
+
+  BOOST_AUTO_TEST_CASE( getLengthTest )
+  {
+      DynamicStringArray obj;
+      obj.addEntry ( "Ana" );
+      obj.addEntry ( "are" );
+      obj.addEntry ( "mere" );
+      BOOST_CHECK( obj.getLength() == 3 );
```
Comment (nbiro) [inline MariaCraciun1_Task2/BoostTest.cpp:46]: I would test if the length is increased when a new entry is added. This test would pass if length is hardcoded to 3

**Review by maxlig79 — CHANGES_REQUESTED** [review]

File: MariaCraciun1_Task2/BoostTest.cpp [hunk 13]
```diff
@@ -0,0 +1,75 @@
...
+  }
+
+  BOOST_AUTO_TEST_CASE( copyConstructorTest )
+  {
+      DynamicStringArray obj;
+      obj.addEntry( "Ana" );
+      DynamicStringArray objCopy;
+      objCopy = obj;
+      if(obj.getLength() != objCopy.getLength() )
+        BOOST_FAIL( "Copy not done correctly" );
```
Comment (nbiro) [inline MariaCraciun1_Task2/BoostTest.cpp:56]: check also if all entries were copied

File: MariaCraciun1_Task2/DynamicStringArray.cpp [hunk 14]
```diff
@@ -0,0 +1,72 @@
...
+    }
+   
+    void DynamicStringArray::addEntry (const std::string &str ) {
+        std::string *tempArray = new std::string[size+1] ;
+        for(int i = 0; i < size; i++ ) {
+            tempArray[i] = dynamicArray[i] ; 
+        }
+        tempArray[size] = str ;
+        size++ ;
+        dynamicArray = tempArray ;
```
Comment (nbiro) [inline MariaCraciun1_Task2/DynamicStringArray.cpp:35]: memory leak, you need to delete dynamicArray

File: MariaCraciun1_Task2/DynamicStringArray.cpp [hunk 15]
```diff
@@ -0,0 +1,72 @@
...
+            status = true;
+        }
+        std::string *tempArray = new std::string[size-1] ;
+        int j=0 ;
+        for ( int i=0; i<size; i++){
+            if ( dynamicArray[i] != str )
+                tempArray[j] = dynamicArray[i] ;
+                j++ ;
+        }
+        dynamicArray = tempArray ;
```
Comment (nbiro) [inline MariaCraciun1_Task2/DynamicStringArray.cpp:54]: memory leak

File: MariaCraciun1_Task2/DynamicStringArray.cpp [hunk 16]
```diff
@@ -0,0 +1,72 @@
...
+        size-- ;
+
+        return status ;
+    }
+
+    std::string *DynamicStringArray::getEntry (const int &index) {
```
Comment (nbiro) [inline MariaCraciun1_Task2/DynamicStringArray.cpp:60]: passing const int by reference doesn't really make sense. Usually, non-trivial types to be passed by reference where copy construction is expensive

**Review by nbiro — CHANGES_REQUESTED** [review]
please don't delete former tasks....

File: MariaCraciun1_Task2/DynamicStringArray.cpp (hunk 16 above)
Comment (MariaCraciun1) [reply MariaCraciun1_Task2/DynamicStringArray.cpp:60]: Point taken. But I don't see the harm in leaving it like this. Please let me know if there is something I am missing.

Comment (nbiro) [reply MariaCraciun1_Task2/DynamicStringArray.cpp:60]: well, under the hood it does harm a little bit, some extra instructions(assuming the compiler is not smart enough to ignore it). more info: https://www.generacodice.com/en/articolo/1258791/C%2B%2B%3A-Why-pass-by-value-is-generally-more-efficient-than-pass-by-reference-for-built-in-(i.e.%2C-C-like)-types

**Pull request closed without merging**